lib/context.js

```diff
@@ -17,30 +17,36 @@ const createContext = function (component, ctxConfig) {
         ctx[ctxConfig.namespace] = {};
       }
 
-      Object.assign(ctx[ctxConfig.namespace], await ctxConfig.factory.call(component, arg)); 
+      Object.assign(ctx[ctxConfig.namespace], await ctxConfig.factory.call(component, context));
     }
 
     return ctx;
   };
 };
 
-const wrapContext = function (contextFunction) {
+const wrapContext = function (component) {
   const middleware = [];
+  const contextFunction = component._context;
+  const dataSourceInject = component._dataSourceInjection;
 
-  const context = async function (arg) {
-    debug('building root context');
+  const context = async function (context) {
+    debug(`building ${component._id} root context`);
 
     for (let { name, fn } of middleware) {
       debug(`applying ${name} middleware`);
-      arg = await fn(arg);
+      context = await fn(context);
     }
 
-    const ctx = await contextFunction(arg);
+    const componentContext = await contextFunction(context);
 
-    return {
-      ...arg,
-      ...ctx
+    const globalContext = {
+      ...context,
+      ...componentContext
     };
+    
+    globalContext.dataSources = dataSourceInject(globalContext);
```

Comment:
if I'm reading things correctly the `dataSources` key on the `context` object injected into resolvers will be `{}` if no datasources are passed into the GraphQLComponent constructor (and assuming no child components have data sources defined?